File: matlab/problem2.m
clear; close all; clc
%% https://projecteuler.net/problem=2


M =  zeros(1,50,'uint32');
n = 0;
idx = 1;

% measure execution speed.
tic
while fibo(n) < 4000000
    n = n + 1;
    
    if mod(fibo(n), 2) == 0
        M(idx) = fibo(n);
        idx = idx + 1;
    end
end
toc
% Elapsed time is 1.221270 seconds.

% reshape matrix to remove any leftover 0's
M = reshape(M(1:idx), 1, []);

% result
disp(sum(M))


% create a recursive fibonacci function for speed
function res = fibo(x)
    if x == 0
        res = 0;
    elseif x == 1
        res = 1;
    else
        res = fibo(x-1) + fibo(x-2);
    end
end

File: c/problem2.c
#include <stdio.h>

unsigned int Fibonacci( unsigned int x );

/*	https://projecteuler.net/problem=2	*/
int main()
{
	unsigned int n = 0;
	unsigned int res = 0;


	do
	{
		if ( Fibonacci( n ) % 2 == 0 )
		{
			res = res + Fibonacci( n );
		}

		n = n + 1;

	} while( Fibonacci( n ) < 4000000 );

	// result
	printf("%d", res);


	return 0;
}

/*	recursive fibonacci function	*/
unsigned int Fibonacci( unsigned int x )
{
	if ( x == 0 )
	{
		return 0;
	}
	else if ( x == 1 )
	{
		return 1;
	}
	else
	{
		return ( Fibonacci( x-1 ) + Fibonacci( x-2 ) );
	}
}
